chore: upgrade to Electron 13 and Node 14

sfoster1 opened the pull request:
# Overview
This PR updates the js stack to recent versions of
- electron (13, from 6 (!))
- node (14, from 12)
- electron-builder (22.12, from 21.2)
- several other transitive dependencies

This is all driven by the desire to update to electron 13.

Electron 13 has a lot of new stuff and improvements in general (https://www.electronjs.org/blog/electron-13-0) but the reason I made this PR is that it updates the chromium web content renderer stack, and the version it updates to has good support for a [wayland](https://wayland.freedesktop.org/) compositor backend on linux. This is a replacement for x11, and it's a lot faster than x11, and it would be good to use it on the OT3 (and allow people to use it on desktop linux as well).

One interesting side effect: not sure what do about this but this now will correctly build m1-native electron builds.

## Non-infra changes
Electron 13 has [some breaking changes](https://www.electronjs.org/docs/breaking-changes#planned-breaking-api-changes-130) relative to older versions. The ones that affected us were
- some shell APIs changed to async versions, requiring some changes to the app code
- context isolation, which is a strict separation between renderer and shell processes rather than implicit rpc translation across the boundary, is now defaulted to on; I manually turned this off, to match the behavior we currently rely on.

## Risk Assessment

High, especially around native integrations like usb detection.

sfoster1 committed:
chore(app,app-shell): Bump electron to 13

Electron 13 has support for a wayland direct renderer without needing to
use x11 bridging which will be quite useful for OT3. It's also a good
idea to catch up on 7 major versions worth of improvements.

sfoster1 committed:
chore(app-shell): contextIsolation back to false

In electron 12, the contextIsolation feature offered by electron to
prevent several very common vulnerabilities triggerable from
user-controlled content in the renderer was defaulted to true. We should
use it! But for it to do anything useful, the ipc interface between
renderer and main process has to be made much more explicit and exposed
across the isolation layer on a per-method (and with sanitization on the
method inputs) basis, so for now just turn it back off.

sfoster1 committed:
docs

sfoster1 committed:
update usb-detection for builds

sfoster1 committed:
node 14 for all workflows

sfoster1 committed:
missed one

sfoster1 committed:
add libudev for usb-detection

sfoster1 committed:
udev needs to be in more places

sfoster1 committed:
better syntax, more places

sfoster1 committed:
maybe this?

sfoster1 committed:
Update app-shell for electron 13 breaking changes

https://www.electronjs.org/docs/breaking-changes#removed-shellmoveitemtotrash
https://www.electronjs.org/docs/breaking-changes#api-changed-shellopenitem-is-now-shellopenpath

y3rsh commented:
Hi @sfoster1 I have tested this on my M1 mac.  I rebased on latest edge so I had 4.5.0.  I had to use `arch -x86_64` to run `make dist-osx` in app-shell `arch -x86_64 make dist-osx` but it worked and I got a dmg that works and when I run the app I see no issues so far.  If you would like we can investigate the issue running without `arch -x86_64`.
![image](https://user-images.githubusercontent.com/502770/129396780-13a72d88-556c-440b-a0d0-e43468d9c0e6.png)

sfoster1 committed:
bump electron-builder

I think we need this to properly notarize on osx anyway, but as long as
we're doing the electron bump we might as well do e-b too and we'll get
support for m1

sfoster1 commented:
> Hi @sfoster1 I have tested this on my M1 mac. I rebased on latest edge so I had 4.5.0. I had to use `arch -x86_64` to run `make dist-osx` in app-shell `arch -x86_64 make dist-osx` but it worked and I got a dmg that works and when I run the app I see no issues so far. If you would like we can investigate the issue running without `arch -x86_64`.
> ![image](https://user-images.githubusercontent.com/502770/129396780-13a72d88-556c-440b-a0d0-e43468d9c0e6.png)

Just to get it noted in the thread: this started working after the update of electron-builder, right?

y3rsh commented:
@sfoster1 with the branch I can successfully build a working .dmg.   I am not certain the electron-builder upgrade is required but with it in place I have successful `make setup` and in app-shell `make dist-osx`.  I believe my issues were with how I had python, node, and make installed on my machine.  The M1 tools setup is challenging.

sfoster1 committed:
Merge branch 'edge' into electron-13

sfoster1 committed:
change node version in new workflows

sfoster1 committed:
need udev on linux

sfoster1 committed:
Merge branch 'edge' into electron-13

mcous commented on app-shell/src/labware/definitions.ts:
```diff
@@ -75,7 +76,12 @@ export function addLabwareFile(file: string, dir: string): Promise<void> {
+    () => {
+      return fs.promises.unlink(file)
+    }
+  )
```
Can this be made simpler? This `() => Promise.resolve()` isn't doing anything. Also, since we're using `fs-extra` for now (probably past time we drop it and move back to vanilla `fs`), I don't think we need `fs.promises` unless the typechecker is mad about it

```suggestion
  return shell.trashItem(file).catch(() => fs.unlink(file))
```

mcous commented on app-shell/src/labware/__tests__/definitions.test.ts:
```diff
@@ -229,7 +229,9 @@ describe('labware directory utilities', () => {
       const setup = fs.writeJson(filename, { name: 'a' })
 
-      moveItemToTrash.mockReturnValue(false)
+      trashItem.mockImplementation(_ =>
+        Promise.reject(Error('something went wrong'))
+      )
```
```suggestion
      trashItem.mockRejectedValue(Error('something went wrong'))
```

mcous commented on app-shell/src/labware/__tests__/definitions.test.ts:
```diff
@@ -213,14 +213,14 @@ describe('labware directory utilities', () => {
       const filename = path.join(dir, 'foo.json')
 
-      moveItemToTrash.mockReturnValue(true)
+      trashItem.mockImplementation(_ => Promise.resolve())
```
```suggestion
      trashItem.mockResolvedValue()
```

mcous reviewed (commented):
Some minor tweak suggestions to promise handling

sfoster1 committed:
Update app-shell/src/labware/definitions.ts

Co-authored-by: Mike Cousins <mike@opentrons.com>

sfoster1 committed:
Update app-shell/src/labware/__tests__/definitions.test.ts

Co-authored-by: Mike Cousins <mike@opentrons.com>

sfoster1 committed:
Update app-shell/src/labware/__tests__/definitions.test.ts

Co-authored-by: Mike Cousins <mike@opentrons.com>

y3rsh reviewed (approved):
- USB worked - I was prompted for realtek usb driver upgrade then it worked after installing that driver.
- Install worked.
- Uninstall worked.
1. on the branch electron-13
2. make teardown
3. make setup
4. cd app-shell
5. make dist-win
6. use .exe to install
7. see robot and docker robot both connected
8. have real robot not connected with wifi
9. update realtek driver
10. did not stop opentrons app
11. robot on usb shows up
12. can home and turn on lights
![image (1)](https://user-images.githubusercontent.com/502770/133456151-375c944f-63aa-4bb0-87e4-40918046447e.png)

b-cooper reviewed (approved):
Looks good on my end 🥥

sfoster1 committed:
nvmrc: node 14

mcous reviewed (approved):
LGTM

sfoster1 committed:
update before install

sfoster1 committed:
missed some
